### PR TITLE
performance improvements

### PR DESCRIPTION
**Pull request by nsd20463**
Cumulatively these changes speed up my benchmarks by 38%.

### CODE REVIEW

**Comment by matteo-grella** [comment]
Hey @nsd20463 thanks a lot for your PR!

Give us a few days for review but I am very impressed by this speedup!

**Comment by nsd20463** [comment]
If you like these, I found more improvements later, while modifying gopickle into a pickle->json converter, some of which can be backported.

**Comment by marco-nicola** [comment]
Whoa, those ended up being _quite_ a few days... :disappointed_relieved: 

Thanks for this very good PR! :heart: 

Trying to be annoyingly picky, as usual: (1) the new capacities feel a little arbitrary; (2) we might want to add some new/better test cases, in particular covering various parts of the reading mechanism. But here are the immediate counter-arguments: (1) Go's default capacities are equally arbitrary, and at we are now reinforced by profiling some real-world usages (that we didn't have time and energy to do before!); (2) tests are already very minimal and incomplete, so I should just shut up! Let's simply keep these in mind as to-dos.

In short: well done, approved! :clap: :partying_face:
